core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -2,52 +2,166 @@
+
+  /**
+   * Keeping DDF without name
+   */
+  private Cache<UUID, DDF> mDDFCache;
+
+  /**
+   * Keeping DDF that has name in a separate cache, so eviction on DDF without name
+   * does not affect DDF with name
+   */
+  private Cache<UUID, DDF> mDDFwithNameCache;
+
+  private DDFManager mDDFManager;
+
+  public DDFCache(DDFManager manager) {
+
+    mDDFManager = manager;
+
+    /**
+     * Maximum number of DDFs without name allowed in cache, DDFs will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFs = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_IN_CACHE));
+
+    /**
+     * Maximum number of DDFs with name allowed in cache, DDFs with name will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFsWName = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_WITH_NAME_IN_CACHE));
+
+    /**
+     * The number of seconds that the DDF will live in cache since it last used(meaning inserted or accessed)
+     */
+    Long timeToIdle = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.DDF_TIME_TO_IDLE_SECONDS));
+
+    mLog.info(String.format("Maximum number of ddfs in cache %s", maxNumberOfDDFs));
+    mLog.info(String.format("DDF's time to idle = %s seconds", timeToIdle));
+
+    mDDFCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFs).expireAfterAccess(timeToIdle, TimeUnit.SECONDS).
+        recordStats().removalListener(new DDFRemovalListener())
+        .build();
+
+    // Don't do time based eviction for DDF with name
+    mDDFwithNameCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFsWName).recordStats().removalListener(new DDFRemovalListener())
+        .build();
+  }
+
+  private boolean containsDDF(UUID uuid) {
+    return (mDDFCache.asMap().containsKey(uuid) || mDDFwithNameCache.asMap().containsKey(uuid));
+  }
 
-  private Map<UUID, DDF> mDDFs = new ConcurrentHashMap<UUID, DDF>();
   private Map<String, UUID> mUris = new ConcurrentHashMap<String, UUID>();
 
   public void addDDF(DDF ddf) throws DDFException {
-    mDDFs.put(ddf.getUUID(), ddf);
+    if(ddf.getUUID() == null) {
+      throw new DDFException("uuid is null");
+    }
+    Lock lock = uuidLocks.get(ddf.getUUID());
+    try {
+      if(lock.tryLock(120, TimeUnit.SECONDS)) {
+        try {
+          if (ddf.getName() == null) {
+            mDDFCache.put(ddf.getUUID(), ddf);
+          } else {
+            mDDFwithNameCache.put(ddf.getUUID(), ddf);
+          }
+        } finally {
+          lock.unlock();
+        }
+      } else {
+        throw new DDFException("Timeout getting DDF");
+      }
+    } catch (InterruptedException e) {
+
+    }
+  }
+
+  public CacheStats getCacheStats() {
+    return mDDFCache.stats().plus(mDDFwithNameCache.stats());
   }
 
-  public void removeDDF(DDF ddf) throws DDFException {
-    mDDFs.remove(ddf.getUUID());
+  public synchronized void removeDDF(DDF ddf) throws DDFException {
```

Comment:
no need to synchronized here.
`id = ddf.getUUID()
if(id != null){
   mDDFCache.invalidate(id)
   mDDFNameCache.invalide(id)
}
`